compiler/src/iree/compiler/Dialect/LinalgExt/IR/AggregatedOpInterfaceImpl.cpp

```diff
@@ -777,6 +777,22 @@ FailureOr<SmallVector<Value>> Im2colOp::decomposeOperation(OpBuilder &b) {
     }
     kBasis.push_back(size);
   }
+
+  // Check if the reduction dimension has different layouts between input and
+  // filter, if so, transpose the order of (P, Q, C) according to
+  // `inputFilterPerm` encoded in im2col metadata.
+  auto isIdentityPermutation = [](ArrayRef<int64_t> perm) -> bool {
+    return llvm::equal(llvm::seq<int64_t>(0, perm.size()), perm);
+  };
+
+  ArrayRef<int64_t> inputFilterPerm = getInputFilterPerm();
+  if (!isIdentityPermutation(inputFilterPerm)) {
+    SmallVector<OpFoldResult> origKBasis = kBasis;
+    for (auto [newIdx, transIdx] : llvm::enumerate(inputFilterPerm)) {
+      kBasis[newIdx] = origKBasis[transIdx];
+    }
+  }
+
```

Comment:
nit: I think this can be simplified using `applyPermutationToVector`:

https://github.com/llvm/llvm-project/blob/dbeb7c1bbbc90f3d556b9d010d7899311cfc62d6/mlir/include/mlir/Dialect/Utils/IndexingUtils.h#L220-L229

```suggestion
  // Transpose the order of (P, Q, C) according to `inputFilterPerm` encoded in im2col metadata.
  applyPermutationToVector(kBasis, getInputFilterPerm());
```

Reply:
Thanks for sharing this useful utility function. This is a great reduction of codes!